docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,534 @@
+  - Upon successful evidence generation, `oe_get_evidence()` returns the
+  generated evidence in a dynamically allocated buffer. It also has the option
+  to return a set of endorsements in another dynamically allocated buffer.
+  After processing of the evidence and the endorsements, the caller is
+  responsible to free these buffers.
+  - The `flags` parameter is a legacy inherited from the legacy API
+  `oe_get_report()` to indicate SGX local or remote attestation.
+  It's redundant -- overlapping with the `evidence_format_uuid` parameter
+  which identifies evidence formats with globally unique UUIDs.
+- Free a dynamically allocated evidence buffer.
+  - Function `oe_result_t oe_free_evidence(uint8_t* evidence_buffer)`,
+  declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h)
+- Free a dynamically allocated endorsements buffer.
+  - Function `oe_result_t oe_free_endorsements(uint8_t* endorsements_buffer)`,
+  declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h)
+- Unregister an attester plugin.
+  - Function `oe_unregister_attester(oe_attester_t* plugin)`,
+  declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+
+Use cases for evidence verification:
+
+- Retrieve a single SGX verifier plugin.
+  - Function `oe_verifier_t* oe_sgx_plugin_verifier(void)`, declared in header
+  file [openenclave/attestation/sgx/verifier.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/sgx/verifier.h).
+  - It returns a single hardcoded SGX verifier plugin for verification of
+  evidences in SGX ECDSA format.
+- Register a verifier plugin.
+  - Function `oe_result_t oe_register_verifier(oe_verifier_t* plugin, const void* config_data, size_t config_data_size)`, declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+  - The verifier plugin returned by `oe_sgx_plugin_verifier()` can be
+  registered with this function.
+  - See notes for the `oe_register_attester()` API function for explanation of
+  the `config_data` parameter.
+- Verify an evidence, optionally using a set of input endorsements and policies.
+  - Function `oe_result_t oe_verify_evidence(const uint8_t* evidence_buffer, size_t evidence_buffer_size, const uint8_t* endorsements_buffer, size_t endorsements_buffer_size, const oe_policy_t* policies, size_t policies_size, oe_claim_t** claims, size_t* claims_length)`,
+  declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+  - Upon successful verification, `oe_verify_evidence` returns a list of claims
```

Comment:
This seems odd to me.  I would expect an API that would take in an Attestation Result format id, and return an Attestation Result buffer (in the requested format) that can be passed to a relying party.   I would think that the verifier doesn't care what the claims are, only the relying party does, so a separate API would take in an attestation result and be able to give information about claims.   Combining them only makes sense, I think, as an optimization for the special case when the Verifier and Relying Party are in the same entity.   See https://tools.ietf.org/html/draft-ietf-rats-architecture.   Alternatively, if your definition of "verifier" is really what the RATS architecture calls a "Relying Party" then I would suggest the APIs should be renamed to match the standard terminology.

Reply:
As explained in an earlier comment, the OE SDK attestation API function `oe_verify_evidence()` is focused on verification of the input evidence. It outputs the verification result, as a returned error code plus a set of claims contained in the evidence if there is no error. The application that calls this API function implements the complete Verifier role, and is responsible to convert the result into a signed Attestation Result if the Relying Party is in a different security domain.

I do think the built-in base claims in the evidence should be separate from the custom claims. This is because there is no limitation on what names custom claims can have. It's possible for a custom claim to have name conflict a base claim. I'll update the proposal to separate these two types of claims.

Reply:
This PR is experimental and short term. The long term requirements are expressed in PR#2959

Reply:
The names for each of the claims would need to be globally unique. In various formats, names are ints, strings, void etc. 
There should never be a case where custom claims and base claims with the same name. The verifier should just fail if this happens.
For future intent: Instead of name value pairs, use a flat buffer. The relying party parses the buffer.
This is not a blocker for the v0.10 API.